abracadabra/src/writer/persistency_manager.cc

```diff
@@ -0,0 +1,64 @@
+#include "persistency_manager.hh"
+#include "g4-mandatory/trajectory.hh"
+
+
+persistency_manager::persistency_manager() {
+    G4String hdf5file = "test_file.h5";
+    nevt = 0;
+    open_file(hdf5file);
+}
+
+void persistency_manager::open_file(G4String filename) {
+    h5writer = new hdf5_writer();
+    h5writer->open(filename);
+}
+
+
+void persistency_manager::close_file() {
+    h5writer->close();
+}
+
+
+G4bool persistency_manager::Store(const G4Event* event) {
+    G4TrajectoryContainer* trajectories = event->GetTrajectoryContainer();
+    if (trajectories) {
+        store_trajectories(trajectories);
+    }
+
+    nevt++;
+    return true;
+}
+
+
+G4bool persistency_manager::Store(const G4Run* run) {
+    // Store the number of events to be processed
+    auto* app = G4RunManager::GetRunManager();
+    G4int num_events = app->GetNumberOfEventsToBeProcessed();
+    G4String key = "num_events";
+    h5writer->write_run_info(key, std::to_string(num_events).c_str());
+
+
+    return true;
+}
+
+void persistency_manager::store_trajectories(G4TrajectoryContainer* trajectories) {
+    for(int i=0; i<trajectories->size(); i++) {
+        trajectory* trj = dynamic_cast<trajectory*>((*trajectories)[i]);
+        if (!trj) continue;
+
+        G4double      mass    = trj->GetParticleDefinition()->GetPDGMass();
+        G4ThreeVector ini_mom = trj->GetInitialMomentum();
+        G4double      energy  = sqrt(ini_mom.mag2() + mass*mass);
+
+        float kin_energy = energy - mass;
+        char primary = 0;
+        G4int mother_id = 0;
+        if (!trj->GetParentID()) {
+            primary = 1;
+        } else {
+            mother_id = trj->GetParentID();
+        }
+
+        h5writer->write_particle_info(nevt, trj, primary, mother_id, kin_energy);
```

Comment:
`write_particle_info` takes a `trajectory*` and extracts lots of information from it, but

+ `primary`
+ `mother_id`
+ `kin_energy`

are extracted from the trajectory right here, and passed in!

Is there any reason for this inconsistency?


Furthermore, `primary` has declared type `char` but the parameter to which it is passed is `bool`. Is there any logic to this madness?

Reply:
> Is there any reason for this inconsistency?

This can be done in a better way. The idea was that `write_particle_info` should simply receive the data to be written, without the need of doing any (physics) computation: just getting some container with all the information, parsing it in the needed way and store it in the HDF5 file. 

In that way, all the computation is done in `persistency_manager::store_trajectories` and, possibly, different type of writers would need to implement a function to receive the particle information and write it in the corresponding format.

> Furthermore, primary has declared type char but the parameter to which it is passed is bool. Is there any logic to this madness?

I think it is fixed now. In the persistency_manager `primary` should be a bool (it's the logical representation) and inside `hdf5_writer::write_particle_info` has to be casted to `char` since HDF5 does not have a `bool` data type.